File: 0496. Next Greater Element I/0496.py
# time complexity: O(len(nums1) + len(nums2))
# space complexity: O(len(nums2))
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        # dictionary to store next greater element for each num in nums2.
        next_greater = {}
        stack = []
        
        # process nums2 to build the mapping.
        for num in nums2:
            # while stack is not empty and current num is greater than stack's top element.
            while stack and stack[-1] < num:
                next_greater[stack.pop()] = num
            stack.append(num)
        
        # build the result for nums1 using the mapping.
        result = []
        for num in nums1:
            if num in next_greater:
                result.append(next_greater[num])
            else:
                result.append(-1)
        return result
        # return [next_greater.get(num, -1) for num in nums1]

File: 0496. Next Greater Element I/0496.java
// time complexity: O(n*m) in the worst case.
// space complexity: O(1)
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] result = new int[nums1.length];

        // for each element in nums1
        for (int i = 0; i < nums1.length; i++) {
            int k = 1;
            // find the position of nums1[i] in nums2
            for (int j = 0; j < nums2.length; j++) {
                if (nums1[i] == nums2[j]) {
                    // look for the next greater element to the right
                    while (j + k < nums2.length && !(nums1[i] < nums2[j + k])) {
                        k++;
                    }
                    // if not found, set -1; else set the next greater element
                    if (j + k == nums2.length) result[i] = -1;
                    else result[i] = nums2[j + k];
                }
            }
        }

        return result;
    }
}
